### ISSUE TITLE
Am I Fully Connected Indicator [Rough Draft]

### ISSUE THREAD

**Issue by jjeffryes** [OpenBazaar/openbazaar-desktop#1278]
## User Story
When a node can't accept incoming messages, it degrades the OpenBazaar experience. Buyers have to pay an extra fee as if the node is offline, and data from other users on the network does not arrive in real time. As a user, I want to know if my node is unable to accept incoming messages, and what steps I can take to fix that.

## Functional Requirements- [ ] When the app starts, there should be an easy to see visual indicator that the node cannot accept incoming messages.
- [ ] There should be an obvious action the user can take to see an explanation of the visual indicator.
- [ ] There should be an obvious action the user can take to receive information on how to allow incoming messages. This information could be on a website external to the application.
- QUESTION: should there be a "messages can be received" indicator, ie: everything is ok, or should the indicator only appear when messages can't be received?
- QUESTION: should there be an option to not show the indicator again?
- QUESTION: if the indicator can be dismissed, should a version that is always present be added somewhere else, like settings/advanced? 

## Data Requirements
- [ ] There will need to be a call the client can make to determine if the node is accepting incoming messages.

**Comment by christroutner** [comment]
Yes please! This is a great idea.

**Comment by placer14** [comment]
In discussion of your questions:
1. The indicator is for informing of a less-than-ideal experience and to prompt action from the user. Making the indicator only visible when there's something to be done about it seems to make sense. A disconnection icon when hovered might reveal "Node Appears Offline to the Network" or similar. Alternatively, we could add a notification on startup when the node appears offline. The fact that it would appear on every startup may be desired for a traveling node. And further, since notifications are attached to temporal events (incoming message, new order, etc), having a "Node Appears Offline" notification attached to a specific startup might make sense.
2. If it's a notification, we probably need to make it optional so it's not super annoying. If we find a nice place for an icon separate from the notifications which casually shows the state, then the option doesn't matter as much if it doesn't clutter the UI.
3. I like the idea of being able to trigger the test again. In the case of a notification, I might want links baked into the notification that lead me to "Details" about the problem which include instructions for testing again in the Advanced settings.

**Comment by cpacia** [comment]
I'm personally not very hot on having another piece of infrastructure needed make the app run. That plus also having it effectively phone home every startup is a bit of a privacy concern. At minimum this functionality should be disabled with tor running.

And probably should even be optional rather than having it fire on every startup.

**Comment by jjeffryes** [comment]
@cpacia what do you mean by phoning home? Wouldn't the indicator come from a call to the server? Or do we need to call some external 3rd party to do this?

**Comment by placer14** [comment]
@cpacia I understand the point about privacy. Leaning toward making this an opt-in feature just for the privacy concerns alone.

@jjeffryes The server would need to check something externally to tell the node whether it is available from the perspective of the outside. The node itself would not be able to tell.

**Comment by rmisio** [comment]
> At minimum this functionality should be disabled with tor running.

Would this even work on Tor even if someone wanted?

Before it got to the service it would bounce around a bunch of other servers and the ip the service would be checking would not be your node. Is that right?

**Comment by BazaarDog** [comment]
How difficult would it be for a suite of tests to be available to the user in test mode?

**Comment by placer14** [comment]
It could probably be delivered easily as a separate command on the openbazaar-go binary in addition to any UI-facing behavior. I expect the UI would trigger it via an API poll or similar. Wouldn't be hard to additionally trigger via a command as well. Something like `openbazaar-go doctor` which runs a bunch of diagnostics and returns a bunch of suggestions on how to improve your installation? (Not unlike `brew doctor`.)

**Comment by BazaarDog** [comment]
I was thinking of something where the test is called from the client. But a server tool would be great too. 

There is a lot of stuff that can break.  I think it would be very useful to have something that will actually complete test transactions on testnet, that would go through all permutations of moderation, disputes, timeouts, etc. It could give a plain text report of the test results without unique identifying information and a check sum.

If someone can't use open bazaar because they followed a DIY config tutorial, they are going to blame the software. A test or diagnostic tool would save a lot of grief.

**Comment by christroutner** [comment]
Correct me if I'm wrong, but this feature is based on detecting when the user is behind a firewall and thus ports 4001, 4002, and 9005 are not accessible. What are some ways the OB client or server can detect this situation?

**Comment by placer14** [comment]
You're right. We're talking about where might be a good place to keep such a feature. I was fantasizing about `ob doctor` being able to check it's online availability as part of a suite of tests the user could execute and get a report on. Would help self troubleshooting and could be another piece of helpful data to provide for support requests. It would also be a good place for us to collect other debug data we expose in various places into a single location (like Block Data? Peer connectivity? Outstanding publish request pending? Etc)

**Comment by christroutner** [comment]
I love the idea @placer14.

@cpacia @placer14 My understanding is that at startup the OB server finds nearby peers to communicate with and download updates from. Would it be possible at that point to detect issues with the ports?

I'm trying to get away from the idea that we would need to connect to some centralized point, e.g. 'phoning home'.

**Comment by placer14** [comment]
I am imagining this is a simple service which receives a multihash of the interested node. Depending on the transport used in the multihash (MH), the service responds appropriately to indicate how various other nodes would perceive that node. At first, it could simply indicate whether a connection could be established and report back success or fail.

Further, being "online" or not probably has a lot to do with whether they are connected in Dual Stack mode or Tor only or using the open internet. We might need to represent additionally if external tor nodes are unable to connect to the "online" node or not as well. Someone might be "online" for the open web but not for tor and that might be important for the user.

Potential service response states:
- Online (available to tor and open web)
- Online for Tor Network Only
- Online for Open Web only (is this even possible.... tor nodes would always be able to reach open web nodes?)
- Offline

**Comment by christroutner** [comment]
I think you could reduce the scope. I think it's a fair assumption that 80%+ of users are loading the OB application on a windows or mac OS, behind a firewall. 

I think it's also a fair assumption that this is the user group that needs this feature most, as they complain the most about fees and transactions not going through (due to non-fully-connected nodes).

Tor users, linux users, etc are important users, but they are not the majority.
